lib/operations/db_ops.js

```diff
@@ -0,0 +1,980 @@
+const MongoError = require('mongodb-core').MongoError;
+const parseIndexOptions = require('../utils').parseIndexOptions;
+const ReadPreference = require('mongodb-core').ReadPreference;
+const toError = require('../utils').toError;
+
+const debugFields = [
+  'authSource',
+  'w',
+  'wtimeout',
+  'j',
+  'native_parser',
+  'forceServerObjectId',
+  'serializeFunctions',
+  'raw',
+  'promoteLongs',
+  'promoteValues',
+  'promoteBuffers',
+  'bufferMaxEntries',
+  'numberOfRetries',
+  'retryMiliSeconds',
+  'readPreference',
+  'pkFactory',
+  'parentDb',
+  'promiseLibrary',
+  'noListener'
+];
+
+// Filter out any write concern options
+const illegalCommandFields = [
+  'w',
+  'wtimeout',
+  'j',
+  'fsync',
+  'autoIndexId',
+  'strict',
+  'serializeFunctions',
+  'pkFactory',
+  'raw',
+  'readPreference',
+  'session'
+];
+
+/**
+ * Add a user to the database.
+ * @method
+ * @param {Db} db The Db instance on which to add a user.
+ * @param {string} username The username.
+ * @param {string} password The password.
+ * @param {object} [options=null] Optional settings. See Db.prototype.addUser for a list of options.
+ * @param {Db~resultCallback} [callback] The command result callback
+ */
+function addUser(db, username, password, options, callback) {
+  // Did the user destroy the topology
+  if (db.serverConfig && db.serverConfig.isDestroyed())
+    return callback(new MongoError('topology was destroyed'));
+  // Attempt to execute auth command
+  _executeAuthCreateUserCommand(db, username, password, options, (err, r) => {
+    // We need to perform the backward compatible insert operation
+    if (err && err.code === -5000) {
+      const finalOptions = applyWriteConcern(Object.assign({}, options), { db: db }, options);
+
+      // Use node md5 generator
+      const md5 = crypto.createHash('md5');
+      // Generate keys used for authentication
+      md5.update(username + ':mongo:' + password);
+      const userPassword = md5.digest('hex');
+
+      // If we have another db set
+      const db = options.dbName ? new Db(options.dbName, db.s.topology, db.s.options) : db;
+
+      // Fetch a user collection
+      const collection = db.collection(Db.SYSTEM_USER_COLLECTION);
+
+      // Check if we are inserting the first user
+      collection.count({}, finalOptions, (err, count) => {
+        // We got an error (f.ex not authorized)
+        if (err != null) return handleCallback(callback, err, null);
+        // Check if the user exists and update i
+        collection
+          .find({ user: username }, { dbName: options['dbName'] }, finalOptions)
+          .toArray(err => {
+            // We got an error (f.ex not authorized)
+            if (err != null) return handleCallback(callback, err, null);
+            // Add command keys
+            finalOptions.upsert = true;
+
+            // We have a user, let's update the password or upsert if not
+            collection.update(
```

Comment:
you can see, by the time we've gotten here we've actually called `executeOperation` _three_ times!